controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller_intg_test.go

```diff
@@ -0,0 +1,227 @@
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+	vmopv1common "github.com/vmware-tanzu/vm-operator/api/v1alpha3/common"
+	"github.com/vmware-tanzu/vm-operator/pkg/constants/testlabels"
+
+	"github.com/vmware-tanzu/vm-operator/test/builder"
+)
+
+const (
+	finalizerName = "virtualmachinereplicaset.vmoperator.vmware.com"
+)
+
+func intgTests() {
+	Describe(
+		"Reconcile",
+		Label(
+			testlabels.Controller,
+			testlabels.EnvTest,
+			testlabels.V1Alpha3,
+		),
+		intgTestsReconcile,
+	)
+}
+
+func intgTestsReconcile() {
+
+	var (
+		ctx *builder.IntegrationTestContext
+
+		rs    *vmopv1.VirtualMachineReplicaSet
+		rsKey types.NamespacedName
+	)
+
+	BeforeEach(func() {
+		ctx = suite.NewIntegrationTestContext()
+		rs = &vmopv1.VirtualMachineReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dummy-replicaset",
+				Namespace: ctx.Namespace,
+			},
+			Spec: vmopv1.VirtualMachineReplicaSetSpec{
+				Replicas: ptr.To(int32(2)),
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"appname": "db",
+					},
+				},
+				Template: vmopv1.VirtualMachineTemplateSpec{
+					ObjectMeta: vmopv1common.ObjectMeta{
+						Labels: map[string]string{
+							"appname": "db",
+						},
+						Annotations: make(map[string]string),
+					},
+					Spec: vmopv1.VirtualMachineSpec{
+						ImageName:  "dummy-image",
+						ClassName:  "dummy-class",
+						PowerState: vmopv1.VirtualMachinePowerStateOn,
+						Network: &vmopv1.VirtualMachineNetworkSpec{
+							Interfaces: []vmopv1.VirtualMachineNetworkInterfaceSpec{
+								{
+									Name: "eth0",
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		rsKey = types.NamespacedName{Name: rs.Name, Namespace: rs.Namespace}
+	})
+
+	AfterEach(func() {
+		ctx.AfterEach()
+		ctx = nil
+		intgFakeVMProvider.Reset()
+	})
+
+	getVirtualMachineReplicaSet := func(ctx *builder.IntegrationTestContext, objKey types.NamespacedName) *vmopv1.VirtualMachineReplicaSet {
+		rs := &vmopv1.VirtualMachineReplicaSet{}
+		if err := ctx.Client.Get(ctx, objKey, rs); err != nil {
+			return nil
+		}
+		return rs
+	}
+
+	waitForReplicaSetFinalizer := func(ctx *builder.IntegrationTestContext, objKey types.NamespacedName) {
+		EventuallyWithOffset(1, func() []string {
+			if rs := getVirtualMachineReplicaSet(ctx, objKey); rs != nil {
+				return rs.GetFinalizers()
+			}
+			return nil
+		}).Should(ContainElement(finalizerName), "waiting for VirtualMachineReplicaSet finalizer")
+	}
+
+	ensureReplicas := func(ctx *builder.IntegrationTestContext, labels map[string]string, desiredReplicas int) {
+		EventuallyWithOffset(1, func() bool {
```

Comment:
Using https://onsi.github.io/gomega/#category-3-making-assertions-eminem-the-function-passed-into-codeeventuallycode pattern for these `Eventually`s can make them a lot clearer when the assertion fails